### PR TITLE
#5 Add Orchard.Logging module adding NLog to pipeline

### PR DESCRIPTION
**Pull request by jersiovic**
Issue #5 related to this topic was closed but at the end I didn't find lines in Orchard 2 implementing the solution provided there. So, here is PR doing what there is described.

### CODE REVIEW

**Comment by Jetski5822** [comment]
Yup, check out https://github.com/NLog/NLog.Extensions.Logging/blob/master/src/NLog.Extensions.Logging/AspNetExtensions.cs

**Comment by sebastienros** [comment]
Created an issue to talk about it first: https://github.com/OrchardCMS/Orchard2/issues/300

**Comment by jersiovic** [comment]
In the mean time I have moved NLog configuration to the Startup.cs of the Orchard.Cms.Web project for this PR. Maybe it is useful while #300 is designed.

**Comment by sebastienros** [comment]
Yes, this is better. Following up on the issue.

**Comment by sebastienros** [comment]
Is there a way with NLog to get one log file per tenant?

**Comment by jersiovic** [comment]
I don't know an specific way for doing that. In orchard one I passed a parameter to Log4Net with the tenant name.
In NLog sth like that can be done with this https://github.com/NLog/NLog/wiki/Event-Context-Layout-Renderer but maybe it would be more desirable to use one of those renderers, storing in httpcontext or in request the tenant name https://github.com/NLog/NLog/wiki/Layout-Renderers#nlogweb-package-
Next in NLog config you can configure one file per tenant

**Comment by agriffard** [comment]
Please, can you resolve the conflicts?

**Comment by jersiovic** [comment]
Sorry I din't read you. ASAP I will do it

**Comment by jersiovic** [comment]
Rebase done, you can merge it

**Comment by agriffard** [comment]
I tested it after retrieving your xkproject branch. 
It logged in src\Orchard.Cms.Web\App_Data\Logs\orchard-log-2017-01-20.log

I had multiple Warnings after the setup :
`2017-01-20 12:17:07.9323|0|Orchard.Environment.Shell.ShellFeaturesManager|WARN|Additional features need to be enabled. `

I created a tenant and tried to generate an error (ex: change the site theme to 'sdf' and reload the homepage).
Is the tenant name supposed to appear somewhere?

Do you have examples of actions creating a log?

Can we have more details about the stack trace, file, line number, ...?

**Comment by sebastienros** [comment]
We need the tenant to be rendered, in O1 it's complex to do, how is it with NLog? 
I assume the "line number" is rendered with the message of the exception, so I don't think there is anything special to do.

**Comment by jersiovic** [comment]
Sorry I only updated this branch but I didn't know you want I render tenant.
I've just committed changes for showing tenant and stacktrace in the log

**Comment by jersiovic** [comment]
@agriffard said:
> Do you have examples of actions creating a log?

Do you mean an Nlog cofig file? Sth like that http://www.peteonsoftware.com/index.php/2008/07/29/nlog/ ?
In general any action you can use can be found in the NLog wiki https://github.com/NLog/NLog/wiki . Specially useful is its [Search in documentation](https://github.com/NLog/NLog/search?q=&type=Wikis&utf8=%E2%9C%93) link.

> Can we have more details about the stack trace, file, line number, ...?

For getting more info related with exceptions, stack trace, file and line you can use these renderers:
https://github.com/NLog/NLog/wiki/Stack-Trace-Layout-Renderer
https://github.com/NLog/NLog/wiki/Exception-layout-renderer
https://github.com/NLog/NLog/wiki/Callsite-Layout-Renderer

**Comment by jersiovic** [comment]
No we are not logging RequestID. Sth like that would be ok? https://github.com/nlog/nlog/wiki/Trace-Activity-Id-Layout-Renderer

**Comment by jersiovic** [comment]
Correction what I mentioned doen't work for Net.Core. The easiest way to get that is to use a custom layout renderer based on a lambda that prints TraceIdentifier:
`LayoutRenderer.Register("trace-identifier", (logEvent) => HttpContext.Current.TraceIdentifier);`
Is this equivalet to the RequestID you asked, isn't it?

**Comment by jersiovic** [comment]
Done.
Now tenant name is stored on httpcontext through a middleware declared on Orchard.Cms.Web project and traceidentifier has been added to nlog.config

File: src/Orchard.Cms.Web/StoreTenantNameAtHttpContextMiddleware.cs [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+            RequestDelegate next,
+            IRunningShellTable runningShellTable)
+        {
+            _next = next;
+            _runningShellTable = runningShellTable;
+        }
+
+        public async Task Invoke(HttpContext httpContext)
+        {
+            var shellSetting = _runningShellTable.Match(httpContext);
```
Comment (sebastienros) [inline src/Orchard.Cms.Web/StoreTenantNameAtHttpContextMiddleware.cs]: ShellSettings is already added as a Feature to HttpContext

Comment (sebastienros) [inline src/Orchard.Cms.Web/StoreTenantNameAtHttpContextMiddleware.cs]: Well but I assume you need a property for the logging ...

File: src/Orchard.Cms.Web/Startup.cs [hunk 2]
```diff
@@ -48,7 +48,9 @@ public void Configure(IApplicationBuilder app, IHostingEnvironment env, ILoggerF
                 loggerFactory.AddDebug();
             }
 
-            app.UseModules();
```
Comment (sebastienros) [inline src/Orchard.Cms.Web/Startup.cs]: ???

File: src/Orchard.Cms.Web/Startup.cs [hunk 3]
```diff
@@ -48,7 +48,9 @@ public void Configure(IApplicationBuilder app, IHostingEnvironment env, ILoggerF
...
+            app.UseMiddleware<ModularTenantContainerMiddleware>();
+            app.UseMiddleware<StoreTenantNameAtHttpContextMiddleware>();
```
Comment (jersiovic) [inline src/Orchard.Cms.Web/Startup.cs]: I need the tenant name stored at httpcontext just after the tenant pipeline is initialized to log properly the tenant name on the first request after starting the server. That's why I need to replace `app.UseModules()` per what it does. 
I'm not happy with the solution of not using `app.UseModules()` .
@sebastienros previously I set it within ModularTenantContainerMiddleware but you told me I should avoid to have logging code specific for a log library within OrchardCore. However thinking it twice to store tenant name in httpcontext for sharing it with other middlewares is not specific code for an specific log library. Don't you prefer the previous solution than this one?

File: src/Orchard.Cms.Web/Startup.cs (hunk 2 above)
Comment (jersiovic) [inline src/Orchard.Cms.Web/Startup.cs]: I explained it on next comment.
This is the implementation of `UseModules()`
```
public static IApplicationBuilder UseModules(this IApplicationBuilder app)
        {
            // Ensure the shell tenants are loaded when a request comes in
            // and replaces the current service provider for the tenant's one.
            app.UseMiddleware<ModularTenantContainerMiddleware>();
            app.UseMiddleware<ModularTenantRouterMiddleware>();

            return app;
        }
```
Problem is I need the new middleware registered just after ModularTenantContainerMiddleware registration. So, I replaced UseModules() per its implementation with the new middleware registered in the position I need.

File: src/Orchard.Cms.Web/StoreTenantNameAtHttpContextMiddleware.cs (hunk 1 above)
Comment (jersiovic) [inline src/Orchard.Cms.Web/StoreTenantNameAtHttpContextMiddleware.cs]: ah ok. Let me check if I can access directly to it

**Comment by jersiovic** [comment]
Done. I've replaced the new middleware per an NLog layout renderer that directly uses ShelSettings in  Httpcontext.Features. Now it looks good

**Comment by jersiovic** [comment]
After last commit removing unneeded middleware class IMO this PR is ready to merge

File: src/Orchard.Cms.Web/StoreTenantNameAtHttpContextMiddleware.cs (hunk 1 above)
Comment (jersiovic) [inline src/Orchard.Cms.Web/StoreTenantNameAtHttpContextMiddleware.cs]: I've removed this class, this middleware is don't needed anymore.

**Comment by sebastienros** [comment]
Looks very nice and clean. We could work on the list and order of things to log, but can be done later when people really care, details.

File: src/Orchard.Cms.Web/Startup.cs [hunk 4]
```diff
@@ -41,6 +40,9 @@ public void Configure(IApplicationBuilder app, IHostingEnvironment env, ILoggerF
 
             app.UseStaticFiles();
             loggerFactory.AddConsole(Configuration);
```
Comment (sebastienros) [inline src/Orchard.Cms.Web/Startup.cs]: I assume that here we have some configuration for the ConsoleLogger which is different from the NLog one? OR is it possible to used only NLog even for the console? And default the Console to Error, and keep Warning in the log file.

**Comment by Jetski5822** [comment]
Talked to @sebastienros - I think this should be in a separate project inside of the OrchardCore folder...

in here: https://github.com/OrchardCMS/Orchard2/tree/master/src/OrchardCore

called `Orchard.Logging.NLog`

This is not a module, just a separate package.

move, TenantLayoutRenderer, and the setup there.... then in the Startup.cs file, do .UseOrchardNLog.

If you want, I can do it for you? and do a PR to your PR?

**Comment by Jetski5822** [comment]
Changed my mind :) - Lets take this in, and implement the changes as the next step.

Good work @jersiovic ....

File: src/Orchard.Cms.Web/Startup.cs (hunk 4 above)
Comment (jersiovic) [inline src/Orchard.Cms.Web/Startup.cs]: You are right NLog can be used also for console. I will change that on another PR

**Comment by jersiovic** [comment]
@jetski5832 thank you. Yes I see it is better for reusing it to move that to another project. I will change also this in another PR

**Pull request merged**